### PR TITLE
add dsu solution

### PR DESCRIPTION
**Pull request by derpman23**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which includes the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making them.

### CODE REVIEW

File: solutions/usaco-1159.mdx [hunk 1]
```diff
@@ -137,6 +139,139 @@ public class ConnectingTwoBarns {
 }
 ```
 
-</JavaSection>
+</JavaSection>	
+						
+</LanguageSection>
+
+## Solution 2 - DSU/UF
```
Comment (SansPapyrus683) [inline solutions/usaco-1159.mdx:146]:
```suggestion
## Solution 2 - DSU
```

File: solutions/usaco-1159.mdx [hunk 2]
```diff
@@ -137,6 +139,139 @@ public class ConnectingTwoBarns {
...
+
+**Time Complexity:** $\mathcal{O}\left(T(N+M)\right)$
+
+### Explanation
+We can also find the components of each graph by storing the edges in a Disjoint Set Union/Union Find Structure, and query the elements in O(n) time. Afterward, we can find the answer with a two pointer search.
```
Comment (SansPapyrus683) [inline solutions/usaco-1159.mdx:151]:
```suggestion
We can also find the components of each graph by storing the edges in a Disjoint Set Union, and query the elements in O(n) time. Afterward, we can find the answer with a two pointer search.
```

File: solutions/usaco-1159.mdx [hunk 3]
```diff
@@ -137,6 +139,139 @@ public class ConnectingTwoBarns {
...
+#define sz(x) (int) (x).size()
+
+using pi = pair<int,int>;
+#define f first
+#define s second
+#define mp make_pair
+
+// disjoint set union / union find structure
+struct DSU {
+	int v[100000];
```
Comment (SansPapyrus683) [inline solutions/usaco-1159.mdx:176]: 1. wrap struct in codesnip
2. please use a vector & a proper constructor

File: solutions/usaco-1159.mdx [hunk 4]
```diff
@@ -137,6 +139,139 @@ public class ConnectingTwoBarns {
...
+	}
+};
+
+int t, n, m, l, r;
+
+DSU dsu;
+
+vi cmp[100000];
+
+ll d1[100000], d2[100000], ans;
```
Comment (SansPapyrus683) [inline solutions/usaco-1159.mdx:197]: put these array sizes in constants, or even better make them vectors instead

File: solutions/usaco-1159.mdx [hunk 5]
```diff
@@ -137,6 +139,139 @@ public class ConnectingTwoBarns {
...
+				d2[dsu.get(i)] = min(d2[dsu.get(i)], (ll)abs(i - cmp[dsu.get(n - 1)][r]));
+				// break once the rth component is greater than i
+				if (cmp[dsu.get(n - 1)][r] >= i) break;
+				r++;
+			}
+			// correct for overcounting
+			if (r) r--;
+		}
+
+		ans = 5000000000;
```
Comment (SansPapyrus683) [inline solutions/usaco-1159.mdx:264]: just use INT32_MAX

File: solutions/usaco-1159.mdx [hunk 6]
```diff
@@ -137,6 +139,139 @@ public class ConnectingTwoBarns {
...
+				 * find the minimum distance from field i to the rth component
+				 * in barn 2's graph
+				 */
+				d2[dsu.get(i)] = min(d2[dsu.get(i)], (ll)abs(i - cmp[dsu.get(n - 1)][r]));
+				// break once the rth component is greater than i
+				if (cmp[dsu.get(n - 1)][r] >= i) break;
+				r++;
+			}
+			// correct for overcounting
+			if (r) r--;
```
Comment (SansPapyrus683) [inline solutions/usaco-1159.mdx:261]: braces here and everywhere there isn't any

File: solutions/usaco-1159.mdx [hunk 7]
```diff
@@ -137,6 +139,139 @@ public class ConnectingTwoBarns {
...
+		dsu.init(n);
+
+		for (int i = 0; i < m; i++) {
+			int a, b; cin >> a >> b;
+			// join a and b into the same graph
+			dsu.unite(--a, --b);
+		}
+
+		// print 0 if nodes 1 and n are in the same graph
+		if (dsu.get(0) == dsu.get(n - 1)) { cout << "0\n"; continue; }
```
Comment (SansPapyrus683) [inline solutions/usaco-1159.mdx:215]: expand these braces

File: solutions/usaco-1159.mdx [hunk 8]
```diff
@@ -137,6 +139,139 @@ public class ConnectingTwoBarns {
...
+
+vi cmp[100000];
+
+ll d1[100000], d2[100000], ans;
+
+int main() {
+	ios_base::sync_with_stdio(0); cin.tie(0);
+
+	cin >> t;
+	while (t--) {
```
Comment (SansPapyrus683) [inline solutions/usaco-1159.mdx:203]: (optional) use a for loop

File: solutions/usaco-1159.mdx [hunk 9]
```diff
@@ -137,6 +139,139 @@ public class ConnectingTwoBarns {
...
+		x = get(x), y = get(y);
+		if (x == y) { return 0; }
+		if (v[x] > v[y]) swap(x, y);
+		v[x] += v[y];
+		v[y] = x;
+		return 1;
+	}
+};
+
+int t, n, m, l, r;
```
Comment (SansPapyrus683) [inline solutions/usaco-1159.mdx:191]: no need for these globals

File: solutions/usaco-1159.mdx [hunk 10]
```diff
@@ -137,6 +139,139 @@ public class ConnectingTwoBarns {
...
+
+DSU dsu;
```
Comment (SansPapyrus683) [inline solutions/usaco-1159.mdx:193]: no need for this global either

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

**Comment by nathangong** [comment]
https://github.com/cpinitiative/usaco-guide/pull/2480#pullrequestreview-845924681 i think ryan made this PR but he ended up removing the solutions, so what should happen here?

File: solutions/usaco-1159.mdx (hunk 2 above)
Comment (andrewwangva) [inline solutions/usaco-1159.mdx:151]: Maybe expand on it to talk about cost function/outside of how DSU was used.

**Pull request closed without merging**